homeassistant/components/schluter/climate.py

```diff
@@ -0,0 +1,165 @@
+"""Support for Schluter thermostats."""
+from datetime import timedelta
+import logging
+
+from requests import RequestException
+import voluptuous as vol
+
+from homeassistant.components.climate import PLATFORM_SCHEMA, ClimateDevice
+from homeassistant.components.climate.const import (
+    CURRENT_HVAC_HEAT,
+    CURRENT_HVAC_IDLE,
+    HVAC_MODE_HEAT,
+    SUPPORT_TARGET_TEMPERATURE,
+)
+from homeassistant.const import ATTR_TEMPERATURE, CONF_SCAN_INTERVAL
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from . import DATA_SCHLUTER_API, DATA_SCHLUTER_SESSION, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+SCAN_INTERVAL = timedelta(seconds=5)
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {vol.Optional(CONF_SCAN_INTERVAL): vol.All(vol.Coerce(int), vol.Range(min=1))}
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the Schluter thermostats."""
+    session_id = hass.data[DOMAIN][DATA_SCHLUTER_SESSION]
+    api = hass.data[DOMAIN][DATA_SCHLUTER_API]
+    temp_unit = hass.config.units.temperature_unit
+
+    async def async_update_data():
+        try:
+            return api.get_thermostats(session_id) or []
+        except RequestException as err:
+            raise UpdateFailed(f"Error communicating with Schluter API: {err}")
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="schluter",
+        update_method=async_update_data,
+        update_interval=SCAN_INTERVAL,
+    )
+
+    await coordinator.async_refresh()
+
+    async_add_entities(
+        SchluterThermostat(coordinator, idx, temp_unit, api, session_id)
+        for idx, ent in enumerate(coordinator.data)
```

Comment:
This seems dangerous because if the user would add/remove a thermostat, the index is no longer correct.

I would suggest that inside your async_update_data you turn your data into a dictionary keyed by serial number.

```python
        try:
            thermostats = await hass.async_add_executor_job(api.get_thermostats, session_id)
        except RequestException as err:
            raise UpdateFailed(f"Error communicating with Schluter API: {err}")

		if thermostats is None:
            return {}

        return {thermo.serial_number: thermo for thermo in thermostats}
```

Reply:
Good catch, I agree. I've turned the data into a dictionary and made the necessary modifications in the platform to use serial number as the unique identifier & index.